hadoop-hdds/docs/content/design/truncate.md

```diff
@@ -0,0 +1,146 @@
+  repeated TruncatedBlocksTransaction truncatedBlocksTransactions = 1;
+  required int64 cmdId = 2;
+}
+
+// The truncated blocks which are stored in truncatedBlock.db of scm.
+message TruncatedBlocksTransaction {
+     required int64 txID = 1;
+     required int64 containerID = 2;
+     repeated PartialTruncateBlock blocks = 3;
+     // the retry time of sending truncating command to datanode.
+     required int32 count = 4;
+}
+```
+
+## Datanode
+When Datanode receives `TruncateBlocksCommand`, in the RocksDB, Datanode deletes the chunks which are fully truncated, and updates the chunk length and checksum which is partially truncated. Datanode puts `<TRUNCATING_blockID, newLength>` in RocksDB, and returns succ to SCM.
+
+Datanode starts a background thread to process `<TRUNCATING_blockID, newLength>`.
+If it is `FilePerBlock`, we use `FileChannel.truncate` to truncate file to newLength directly.  If it is `FilePerChunk`, we delete the files of the fully truncated chunks, and use `FileChannel.truncate` to process the partially truncated file.  Then, in RocksDB, Datanode delete `<TRUNCATING_blockid, newLength>`, and put `<TRUNCATED_blockid, newLength>`.
+
+## Client
+When the client reads the key, the client gets blocks of the key from OM, including block length. Then the client gets chunks of each block from Datanode.  The client maybe gets the chunks which should be fully or partially truncated, in this case, the client only reads the block length and ignores the other.
+
+## Multi Client
+### Read When Truncate
+#### Case 1
+
+Client1 gets blocks from OM -> Client2 truncates the same file but datanode has not truncated blocks of the file -> Datanode truncates block of the file->  Client1 reads block from Datanode.  Then Client1 will find inconsistency, for example, the block’s length Client1 obtained from OM is 128M, then Client2 truncates the block to 127M, so Client1 reads only 127M rather than 128M. In this case, Client1 throws an exception.
+#### Case 2
+
+Client1 gets blocks from OM -> Client2 truncates the same file but datanode has not truncated blocks of the file ->  Client1 finishes reading block from Datanode -> Datanode truncates block of the file.  So Client1 can read succ.
+
+Both 2 cases are acceptable, HDFS also does not guarantee the behavior of the read client.
+
+### Truncate When Truncate
+
+When multi clients truncate at the same time, the BUCKET_LOCK can avoid inconsistency.
+
+### Truncate when Append
+
+Append mean ozone client can append some content to the tail of some key. So if Client1 truncates the file but datanode has not truncated blockN -> Client2 appends some content to the tail of blockN -> Datanode truncates blockN, then error happens. To avoid this, OmKeyLocationInfo add a new flag, i.e. toBeTruncated, to mark the block need to be truncated in the future. When Client2 appends, Client2 finds blockN with toBeTruncated flag, Client2 will allocate a new block, and append the content to the new block.
```

Comment:
After toBeTruncated is set, it is hard to reset this flag in architecture of current Ozone